### PR TITLE
Batch Write optimization

### PR DESCRIPTION
**Pull request by birdstorm**

### CODE REVIEW

File: core/src/main/scala/com/pingcap/tispark/write/TiBatchWriteTable.scala [hunk 1]
```diff
@@ -185,60 +187,36 @@ class TiBatchWriteTable(
...
-        }
+      val wrappedEncodedRecordRdd = generateRecordKV(distinctWrappedRowRdd, remove = false)
+      val wrappedEncodedIndexRdds = generateIndexKVs(distinctWrappedRowRdd, remove = false)
+      val wrappedEncodedIndexRdd = unionAll(sc, wrappedEncodedIndexRdds)
+
+      splitTableRegion(wrappedEncodedRecordRdd)
+      splitIndexRegion(wrappedEncodedIndexRdds)
+
+      wrappedEncodedRecordRdd ++ generateRecordKV(deletion, remove = true) ++
+        wrappedEncodedIndexRdd ++ generateIndexKV(sc, deletion, remove = true)
```
[the remark this answers is not in the document]
Comment (birdstorm) [reply core/src/main/scala/com/pingcap/tispark/write/TiBatchWriteTable.scala]: fixed

File: core/src/main/scala/com/pingcap/tispark/write/TiBatchWrite.scala [hunk 2]
```diff
@@ -207,8 +211,10 @@ class TiBatchWrite(
     logger.info(s"primary key: $primaryKey")
 
     // filter primary key
-    val secondaryKeysRDD = shuffledRDD.filter { keyValue =>
-      !keyValue._1.equals(primaryKey)
+    val secondaryKeysRDD = shuffledRDD.mapPartitions { iter =>
+      iter.map { keyValue =>
+        new BytePairWrapper(keyValue._1.bytes, keyValue._2)
+      }
```
[the remark this answers is not in the document]
Comment (birdstorm) [reply core/src/main/scala/com/pingcap/tispark/write/TiBatchWrite.scala]: reverted.

File: core/src/main/scala/com/pingcap/tispark/TiConfigConst.scala [hunk 3]
```diff
@@ -41,6 +41,7 @@ object TiConfigConst {
   val WRITE_ALLOW_SPARK_SQL: String = "spark.tispark.write.allow_spark_sql"
   val WRITE_ENABLE: String = "spark.tispark.write.enable"
   val WRITE_WITHOUT_LOCK_TABLE: String = "spark.tispark.write.without_lock_table"
+  val PREWRITE_CONCURRENCY: String = "spark.tispark.write.prewrite_concurrency"
```
Comment (marsishandsome) [inline core/src/main/scala/com/pingcap/tispark/TiConfigConst.scala:44]: delete this line

File: tikv-client/src/main/java/com/pingcap/tikv/TiConfiguration.java [hunk 4]
```diff
@@ -57,6 +57,7 @@
   private static final int DEF_KV_CLIENT_CONCURRENCY = 10;
   private static final List<TiStoreType> DEF_ISOLATION_READ_ENGINES =
       ImmutableList.of(TiStoreType.TiKV, TiStoreType.TiFlash);
+  private static final int DEF_PREWRITE_CONCURRENCY = 20;
```
Comment (marsishandsome) [inline tikv-client/src/main/java/com/pingcap/tikv/TiConfiguration.java:60]: delete this line

File: tikv-client/src/main/java/com/pingcap/tikv/TiConfiguration.java [hunk 5]
```diff
@@ -84,6 +85,8 @@
 
   private List<TiStoreType> isolationReadEngines = DEF_ISOLATION_READ_ENGINES;
 
+  private int prewriteConcurrency = DEF_PREWRITE_CONCURRENCY;
```
Comment (marsishandsome) [inline tikv-client/src/main/java/com/pingcap/tikv/TiConfiguration.java:88]: delete this line

File: core/src/main/scala/com/pingcap/tispark/write/TiBatchWriteTable.scala [hunk 6]
```diff
@@ -190,60 +197,51 @@ class TiBatchWriteTable(
         throw new TiBatchWriteException("data to be inserted has conflicts with TiKV data")
       }
 
-      val wrappedEncodedRdd = generateKV(distinctWrappedRowRdd, remove = false)
-      splitTableRegion(wrappedEncodedRdd.filter(r => !r.isIndex))
-      splitIndexRegion(wrappedEncodedRdd.filter(r => r.isIndex))
+      val wrappedEncodedRecordRdd = generateRecordKV(distinctWrappedRowRdd, remove = false)
+      val wrappedEncodedIndexRdds = generateIndexKVs(distinctWrappedRowRdd, remove = false)
+      val wrappedEncodedIndexRdd = sc.union(wrappedEncodedIndexRdds.values.toSeq)
```
Comment (marsishandsome) [inline core/src/main/scala/com/pingcap/tispark/write/TiBatchWriteTable.scala:202]:
```suggestion
      val wrappedEncodedIndexRdd: RDD[WrappedEncodedRow] = {
        val list = wrappedEncodedIndexRdds.values.toSeq
        if(list.isEmpty) {
          sc.emptyRDD[WrappedEncodedRow]
        } else if(list.size == 1) {
          list.head
        } else {
          sc.union(list)
        }
      }
```

File: core/src/main/scala/com/pingcap/tispark/write/TiBatchWriteTable.scala [hunk 7]
```diff
@@ -190,60 +197,51 @@ class TiBatchWriteTable(
...
       val wrappedRowRdd = tiRowRdd.zipWithIndex.map { row =>
         WrappedRow(row._1, row._2 + start)
       }
 
-      val wrappedEncodedRdd = generateKV(wrappedRowRdd, remove = false)
-      splitTableRegion(wrappedEncodedRdd.filter(r => !r.isIndex))
-      splitIndexRegion(wrappedEncodedRdd.filter(r => r.isIndex))
+      val wrappedEncodedRecordRdd = generateRecordKV(wrappedRowRdd, remove = false)
+      val wrappedEncodedIndexRdds = generateIndexKVs(wrappedRowRdd, remove = false)
+      val wrappedEncodedIndexRdd = sc.union(wrappedEncodedIndexRdds.values.toSeq)
```
Comment (marsishandsome) [inline core/src/main/scala/com/pingcap/tispark/write/TiBatchWriteTable.scala:232]: val wrappedEncodedIndexRdd: RDD[WrappedEncodedRow] = {
        val list = wrappedEncodedIndexRdds.values.toSeq
        if(list.isEmpty) {
          sc.emptyRDD[WrappedEncodedRow]
        } else if(list.size == 1) {
          list.head
        } else {
          sc.union(list)
        }
      }

File: core/src/main/scala/com/pingcap/tispark/write/TiBatchWriteTable.scala [hunk 8]
```diff
@@ -812,28 +892,30 @@ class TiBatchWriteTable(
...
-          value != null && value.toString != null
-        }
+        val rdd = wrappedEncodedRdd(index.getId)
+
         val regionSplitNum = if (options.regionSplitNum != 0) {
           options.regionSplitNum
         } else {
-          estimateRegionSplitNum(rdd) * 4
+          val indexSize = calcSize(rdd)
+          logger.info(s"count=$count index=$indexSize")
```
Comment (marsishandsome) [inline core/src/main/scala/com/pingcap/tispark/write/TiBatchWriteTable.scala:901]:
```suggestion
          logger.info(s"count=$count indexSize=$indexSize")
```

**Pull request merged**